Add document describes how a get/commit is done in FDB (#4507)
apple/foundationdb: design/Commit/How a commit is done in FDB.md

  * Otherwise return `Never`

  * Increase current commit version, return it back to the commit proxy.

    * Only one process serves as master. Thus the commit version is unique for each cluster.

    * The monotonically increasing commit version will ensure each transaction processed in strict ordering.

### Resolution section

* The commit proxy sends the transaction to the resolver.
* Resolver waits until its version reaches `prevVersion`
  * Ensures all transactions having version smaller than this transaction are resolved.
  * Detects conflicts for the given transaction:
    * If there is no conflict, return `TransactionCommitted` as the status
    * Any conflict, return `TransactionConflict` status
    * If the read snapshot is not in MVCC, return `TransactionTooOld` status

### Post Resolution section

* The proxy waits until the local batch number is current
* The proxy will update the metadata keys and calculate which storage servers are affected
* The proxy then waits until the commit version is current, i.e. only those commits in the MVCC window should be processed.
* The proxy pushs the commit data to TLog

> * TLog waits the commit version to current, then persist the commit.

Author: Yeah, that makes sense.

### TLog section

* Wait until *all* TLogs returns the transaction result.

### Reply section

* The proxy will update the master its commit version
* Reply the result to the client, base on the result from the resolver.

## Tracking the process using `g_traceBatch`

> `g_traceBatch` can be used for querying the transactions and commits. A typical query string:

Reviewer: Do we want to include Splunk query strings in our documentation? In particular, this includes an `index` field that is definitely not general.

> `g_traceBatch` can be used for querying the transactions and commits. A typical query string:

Reviewer 2: I think it's a good idea to have some sample queries to show how it's been used. I agree if a field is not generally used, then these samples shouldn't include it.

> `g_traceBatch` can be used for querying the transactions and commits. A typical query string:

Reviewer: I think it's useful to describe how one might look at logs, and using Splunk syntax might be an ok choice (this is a more relevant question if we start using various Splunk functions), but I think it's a question worth discussing if we're going to start introducing these things.

I do think we should avoid anything like index names or fields that are specific to one company's implementation.

# How a commit is done in FDB

> ## Overall description

Reviewer 3: Please add a sentence:
This doc describes how commit is done in FDB 6.3+. 
The commit path in FDB 6.3 and before is documented in documentation/sphinx/source/read-write-path.rst (https://github.com/apple/foundationdb/pull/4099)

  * Increase current commit version, return it back to the commit proxy.

    * Only one process serves as master. Thus the commit version is unique for each cluster.

    * The monotonically increasing commit version will ensure that each transaction is processed in a strict serial order.

### Resolution section

* The commit proxy sends the transaction to the resolver.
* Resolver waits until its version reaches `prevVersion`
  * Ensures all transactions having version smaller than this transaction are resolved.
  * Detects conflicts for the given transaction:
    * If there is no conflict, return `TransactionCommitted` as the status
    * Any conflict, return `TransactionConflict` status
    * If the read snapshot is not in MVCC, return `TransactionTooOld` status

### Post Resolution section

* The proxy waits until the local batch number is current

### TLog section
* The proxy updates the metadata keys and attaches corresponding storage servers' tags to all mutations.
* The proxy then waits until the commit version is current, i.e. the proxy's committed version is catching up with the commit version of the batch and these two versions are within the MVCC window.

> * The proxy pushs the commit data to TLog

Reviewer 2: Sorry, my original comment was the about moving the next line to this section. I think these three lines should belong to proxy's post resolution section.

Author: I see, I guess I can remove the TLog section in this case.

> * The proxy pushs the commit data to TLog

Reviewer 2: ```suggestion
* The proxy pushes the commit data to TLogs.
```

> * TLog waits the commit version to current, then persist the commit.

Reviewer 2: ```suggestion
* TLog waits the commit version to be current, then persists the commit.
```

> * Wait until *all* TLogs returns the transaction result.

Reviewer 2: ```suggestion
* Wait until *all* TLogs return the transaction result.
```

### Reply section

* The proxy updates the master with the committed version for next GRV request at the master.
* Reply the result to the client, base on the result from the resolver.

## Tracking the process using `g_traceBatch`

> `g_traceBatch` can be used for querying the transactions and commits. A typical query string for Splunk is:

Reviewer: I know it was proposed that "for Splunk" be added, but I think this looks sufficiently general that we could just write the following:

```suggestion
`g_traceBatch` can be used for querying the transactions and commits. A typical query in the trace logs is:
```

```

> LogGroup=loggroup Type=type Location=location

Reviewer: ```suggestion
Type=<Type> Location=<Location>
```

Reviewer: I'm proposing omitting the `LogGroup` from the query because it's not related to this particular lookup.

```

The format of `location` is, in general, `<source_file_name>.<function/actor name>.<log information>`, e.g.

```
NativeAPI.getConsistentReadVersion.Before
```

means the `location` is at `NativeAPI.actor.cpp`, `ACTOR` `getConsistentReadVersion`, `Before` requesting the read version from GRV Proxy.

> Some of example queries are:

Reviewer: ```suggestion
Some example queries are:
```

```

> LogGroup=loggroup Type=TransactionDebug Location=NativeAPI*

Reviewer: ```suggestion
Type=TransactionDebug Location=NativeAPI*
```

```

```

> LogGroup=loggroup Type=CommitDebug Location=storageserver*

Reviewer: I don't think there are any `CommitDebug` messages on the storage server.

Author: Yeah... Should correct this.
